### PR TITLE
Exposing callbacks to sparseautoencoder

### PR DESCRIPTION
**Pull request by jeffin143**
Handles #2013 

Upon merging this, That issue can be closed :)
The list was growing

### CODE REVIEW

File: src/mlpack/methods/sparse_autoencoder/sparse_autoencoder.hpp [hunk 1]
```diff
@@ -86,6 +86,31 @@ class SparseAutoencoder
...
+   * @tparam CallbackTypes Types of Callback Functions.
+   * @param data Input data with each column as one example.
+   * @param visibleSize Size of input vector expected at the visible layer.
+   * @param hiddenSize Size of input vector expected at the hidden layer.
+   * @param lambda L2-regularization parameter.
+   * @param beta KL divergence parameter.
+   * @param rho Sparsity parameter.
+   */
+  template<typename OptimizerType, typename... CallbackTypes>
+  SparseAutoencoder(const arma::mat& data,
```
Comment (zoq) [inline src/mlpack/methods/sparse_autoencoder/sparse_autoencoder.hpp]: Can you add a description for the optimizer and the callbacks? Also looks like the method above is missing a optimizer description as well, do you mind to add one there as well?

Comment (jeffin143) [reply src/mlpack/methods/sparse_autoencoder/sparse_autoencoder.hpp]: I have added it accordingly

**Comment by zoq** [comment]
Looks like there is so problem with the input data in the callbacks test, let me know if you need any help to figure this out.

**Comment by jeffin143** [comment]
> Looks like there is so problem with the input data in the callbacks test, let me know if you need any help to figure this out.

Can you help me figure it out please, I am not sure of it.

File: src/mlpack/tests/callback_test.cpp [hunk 2]
```diff
@@ -257,4 +259,19 @@ BOOST_AUTO_TEST_CASE(RBMCallbackTest)
...
+  // Simple fake dataset.
+  arma::mat data1("0.1 0.2 0.3 0.4 0.5;"
+                  "0.1 0.2 0.3 0.4 0.5;"
+                  "0.1 0.2 0.3 0.4 0.5;"
+                  "0.1 0.2 0.3 0.4 0.5;"
+                  "0.1 0.2 0.3 0.4 0.5");
+
+  ens::L_BFGS optimizer(5, 100);
+  ens::StoreBestCoordinates<arma::mat> cb;
+  mlpack::nn::SparseAutoencoder encoder2(data1, 2, 1, 0, 0, 0 , optimizer, cb);
```
Comment (zoq) [inline src/mlpack/tests/callback_test.cpp:273]:
```suggestion
  mlpack::nn::SparseAutoencoder encoder2(data1, 5, 1, 0, 0, 0 , optimizer, cb);
```

I think that should work.

Comment (jeffin143) [reply src/mlpack/tests/callback_test.cpp]: Nice find , thanks :)

Comment (zoq) [reply src/mlpack/tests/callback_test.cpp]: Haven't tested it, but let's see.

**Comment by jeffin143** [comment]
@zoq , Test failing are irrelevant to this pr and hence I guess the issue is solved :)

File: src/mlpack/tests/callback_test.cpp [hunk 3]
```diff
@@ -257,4 +259,19 @@ BOOST_AUTO_TEST_CASE(RBMCallbackTest)
   BOOST_REQUIRE_GT(stream.str().length(), 0);
 }
 
+BOOST_AUTO_TEST_CASE(SparseAutoencodeCallbackTest)
```
Comment (zoq) [inline src/mlpack/tests/callback_test.cpp:262]:
```suggestion
/**
 * Tests the SparseAutoencoder implementation with
 * StoreBestCoordinates callback.
 */
BOOST_AUTO_TEST_CASE(SparseAutoencodeCallbackTest)
```

**Comment by zoq** [comment]
I have one more last comment, can you update `HISTORY.md` file?

**Comment by zoq** [comment]
Thanks for another great contribution!

**Pull request merged**